### * 라이브러리 불러오기

In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt

from scipy import stats
from statsmodels.stats.proportion import proportions_ztest
%matplotlib inline
import os
# 작업 디렉터리를 설정
os.chdir('C:\/Users/gksxk/statistics')

### Q1. 
#### 자동차 연비 DataSet에서 기술통계치 구하기

In [12]:
mycars = pd.read_csv("../실습파일/BigData분석기초/mycars.csv", encoding = 'euc-kr')
mycars.head()

manufacturer model  displacement  year  cylinder automatic driving  mpg  \
0         audi    a4           1.8  1999         4      auto       f   18   
1         audi    a4           1.8  1999         4    manual       f   21   
2         audi    a4           2.0  2008         4    manual       f   20   
3         audi    a4           2.0  2008         4      auto       f   21   
4         audi    a4           2.8  1999         6      auto       f   16   

   highway_mileage fuel    class  
0               29    p  compact  
1               29    p  compact  
2               31    p  compact  
3               30    p  compact  
4               26    p  compact

In [14]:
df_group = mycars[['mpg', 'highway_mileage', 'driving']]

##### * 시내에서의 연비(mpg) 통계치

In [15]:
df_group.groupby('driving')['mpg'].describe()

count       mean       std   min   25%   50%   75%   max
driving                                                          
4        103.0  14.330097  2.874459   9.0  13.0  14.0  16.0  21.0
f        106.0  19.971698  3.626510  11.0  18.0  19.0  21.0  35.0
r         25.0  14.080000  2.215852  11.0  12.0  15.0  15.0  18.0

##### * 고속도로에서의 연비(highway_mileage) 통계치

In [17]:
df_group.groupby('driving')['highway_mileage'].describe()

count       mean       std   min   25%   50%   75%   max
driving                                                          
4        103.0  19.174757  4.078704  12.0  17.0  18.0  22.0  28.0
f        106.0  28.160377  4.206881  17.0  26.0  28.0  29.0  44.0
r         25.0  21.000000  3.662877  15.0  17.0  21.0  24.0  26.0

시내에서의 연비(mpg), 고속도로에서의 연비(highway_milege) 모두 'f'에서 가장 높은 평균을 나타낸다.  
반면에 연비의 효율성 면에서는 시내에 대해서는 'r'이 가장 좋은 것으로 보이고, 고속도로에서는 '4'가 제일 좋은 것으로 판단된다.  
다만 시내의 경우 'r'의 케이스가 적어, 케이스가 충분하면서 평균이 'r'과 매우 근소한 차이를 보이는 '4'의 연비 효율이 좋은 것으로도 판단될 수 있다.

### Q2. 
#### 범주형 변수인 구동방식, 모델에 대해 counts, percents, cumulative counts를 계산하기

In [2]:
mycars = pd.read_csv("../실습파일/BigData분석기초/mycars.csv", encoding = 'euc-kr')
mycars.head()

manufacturer model  displacement  year  cylinder automatic driving  mpg  \
0         audi    a4           1.8  1999         4      auto       f   18   
1         audi    a4           1.8  1999         4    manual       f   21   
2         audi    a4           2.0  2008         4    manual       f   20   
3         audi    a4           2.0  2008         4      auto       f   21   
4         audi    a4           2.8  1999         6      auto       f   16   

   highway_mileage fuel    class  
0               29    p  compact  
1               29    p  compact  
2               31    p  compact  
3               30    p  compact  
4               26    p  compact

In [102]:
model_auto = mycars[['model', 'automatic']]
count= model_auto[['model', 'automatic']].value_counts().sort_index()
cumcnt = np.cumsum(count)
percent = ma / sum(count) * 100
cumpct = np.cumsum(percent)
# DataFrame으로 취합
df = pd.DataFrame({'Count': count, 
                         'CumCnt': cumcnt, 
                         'Percent': percent,
                         'CumPct': cumpct})
df.round(3).sort_values('Count', ascending = False).head(60)

Count  CumCnt  Percent   CumPct
model                  automatic                                 
caravan 2wd            auto          11      60    4.701   25.641
grand cherokee 4wd     auto           8     125    3.419   53.419
durango 4wd            auto           7      95    2.991   40.598
ram 1500 pickup 4wd    auto           6     205    2.564   87.607
dakota pickup 4wd      auto           6      85    2.564   36.325
malibu                 auto           5     163    2.137   69.658
civic                  manual         5      69    2.137   29.487
f150 pickup 4wd        auto           5     109    2.137   46.581
mustang                manual         5     179    2.137   76.496
c1500 suburban 2wd     auto           5      35    2.137   14.957
grand prix             auto           5     130    2.137   55.556
explorer 4wd           auto           5     103    2.137   44.017
jetta                  manual         5     152    2.137   64.957
impreza awd            auto           4     139    1.709   59.402
                       manual         4     143    1.709   61.111
jetta                  auto           4     147    1.709   62.821
sonata                 auto           4     217    1.709   92.735
tiburon                manual         4     227    1.709   97.009
range rover            auto           4     213    1.709   91.026
mountaineer 4wd        auto           4     170    1.709   72.650
mustang                auto           4     174    1.709   74.359
passat                 auto           4     192    1.709   82.051
ram 1500 pickup 4wd    manual         4     209    1.709   89.316
k1500 tahoe 4wd        auto           4     156    1.709   66.667
4runner 4wd            auto           4       4    1.709    1.709
toyota tacoma 4wd      manual         4     234    1.709  100.000
a4                     auto           4      10    1.709    4.274
camry solara           auto           4      46    1.709   19.658
a4 quattro             auto           4      17    1.709    7.265
                       manual         4      21    1.709    8.974
camry                  auto           4      39    1.709   16.667
civic                  auto           4      64    1.709   27.350
altima                 auto           3      27    1.282   11.538
navigator 2wd          auto           3     182    1.282   77.778
new beetle             auto           3     185    1.282   79.060
                       manual         3     188    1.282   80.342
a6 quattro             auto           3      24    1.282   10.256
passat                 manual         3     195    1.282   83.333
pathfinder 4wd         auto           3     198    1.282   84.615
a4                     manual         3      13    1.282    5.556
sonata                 manual         3     220    1.282   94.017
tiburon                auto           3     223    1.282   95.299
toyota tacoma 4wd      auto           3     230    1.282   98.291
altima                 manual         3      30    1.282   12.821
expedition 2wd         auto           3      98    1.282   41.880
gti                    manual         3     135    1.282   57.692
corvette               manual         3      79    1.282   33.761
dakota pickup 4wd      manual         3      88    1.282   37.607
forester awd           auto           3     114    1.282   48.718
                       manual         3     117    1.282   50.000
camry                  manual         3      42    1.282   17.949
corolla                auto           3      72    1.282   30.769
camry solara           manual         3      49    1.282   20.940
corolla                manual         2      74    0.855   31.624
corvette               auto           2      76    0.855   32.479
gti                    auto           2     132    0.855   56.410
4runner 4wd            manual         2       6    0.855    2.564
land cruiser wagon 4wd auto           2     158    0.855   67.521
f150 pickup 4wd        manual         2     111    0.855   47.436
maxima  

위의 결과를 토대로, caravan 2wd 모델이면서, auto 형태의 차량이 가장 많은 수가 조사된 것으로 나타났다.  
특히 auto 형태의 차량이 빈도 수가 많은 것으로 나타난 것으로 보아, 운전자들이 주로 차량의 auto 형태를 선호함을 예측해 볼 수 있다.

### Q3. 
#### 대한민국 남성의 20~30대 키를 측정한 결과, 평균 175cm, 표준편차 5cm이다. 나의 키가 180cm일 경우에 몇 % 범위에 위치하는가?

In [18]:
mu = 175
sigma = 5
x = 180
prob = stats.norm.cdf(x, mu, sigma)
print("{0}이상 일 경우, 전체 중 비율 : {1:.1f}%".format(x, (1-prob)*100))

180이상 일 경우, 전체 중 비율 : 15.9%


### Q4. 
#### 어떤 자동차 1리터당 주행거리가 평균 12km, 표준편차 3km인 정규분포를 따른다.

##### 1) 1리터를 가지고 9km 이상 15km 이하를 달릴 확률은 얼마인가?  

In [38]:
mu = 12
sigma = 3
x1 = 9
x2 = 15
prob1 = stats.norm.cdf(x1, mu, sigma)
prob2 = stats.norm.cdf(x2, mu, sigma)
print("1리터를 가지고 {0}km 이상 {1}km 이하를 달릴 확률은 {2:.3f}입니다.".format(x1, x2, (prob2-prob1)))

1리터를 가지고 9km 이상 15km 이하를 달릴 확률은 0.683입니다.


##### 2) 1리터를 가지고 15km 이상을 달릴 확률은 얼마인가?

In [21]:
print("1리터를 가지고 {0}km 이상을 달릴 확률은 {1:.3f}입니다.".format(x2, (1-prob2)))

1리터를 가지고 15km 이상을 달릴 확률은 0.159입니다.


### Q5. 
#### 확률변수 t는 자유도가 100인 t-분포를 따른다. t값이 2.0일 때, P(T<=t)을 계산하시오.

In [30]:
t=2.0
df = 100
prob = stats.t.cdf(t, df).round(3)
print("P(T <= {}) : {:.3f}".format(t, prob))

P(T <= 2.0) : 0.976


### Q6. 
#### 확률변수 카이제곱은 자유도가 8인 카이제곱 분포를 따른다. 카이제곱 값이 3.0일 때, P(X <= X^2(카이제곱))을 계산하시오.

In [32]:
chisq = 3.0
df = 8
prob = stats.chi2.cdf(chisq, df).round(3)
print("P(X <= {0}) : {1:.3f}".format(chisq, prob))

P(X <= 3.0) : 0.066


### Q7. 
#### 확률변수 F는 각각 자유도가 10, 100인 F-분포를 따른다. F값이 1.5 일 때, P(X <= F)을 계산하시오.

In [33]:
f = 1.5
dfnum = 10
dfden = 100
prob = stats.f.cdf(f, dfn=dfnum, dfd=dfden)
print("P(X <= {0}): {1:.3f}".format(f, prob))

P(X <= 1.5): 0.850


### Q8. 
#### 도장공정에서 광택도 불량이 30%나 된다고 한다.  5대의 차량을 임의로 선택했을 때, 불량대수가 각각 0,1,2,3,4,5 대가 나올 확률은 얼마인가?

In [36]:
n = 5
for i in range(n+1):
    prob = stats.binom.pmf(k=i, n=n, p=0.3)
    print("P(X = {0}) = {1:.3f}".format(i, prob))

P(X = 0) = 0.168
P(X = 1) = 0.360
P(X = 2) = 0.309
P(X = 3) = 0.132
P(X = 4) = 0.028
P(X = 5) = 0.002


1대가 불량으로 나올 확률이 0.360으로 가장 높았으며, 5대 모두가 불량으로 나올 확률은 0.002로 매우 낮음을 확인할 수 있다.

### Q9. 
#### 정규성 검정
1) 실습 파일 2sample_t_test.csv를 불러오고  
2) 변수, Mat_A와 Mat_b에 대해 각각 정규성을 검정한다.

In [51]:
df = pd.read_csv("../실습파일/BigData분석기초/2sample_t_test.csv")
df.head()

Mat_A  Mat_B
0   73.4   68.7
1   77.0   71.4
2   73.7   69.8
3   73.3   75.3
4   73.1   71.3

##### Mat_A에 대한 정규성 검정  : 정규성 만족
  
정규성 검정의 가설설정은 다음과 같다.  
H0 : 정규성을 만족한다.  
H1 : 정규성을 만족하지 못한다.  
  
p-value가 0.256으로 0.05보다 크므로, 귀무가설을 기각하기 어렵다.  
따라서 Mat_A의 분포는 정규성을 만족한다고 할 근거가 충분하다.

In [52]:
statistic_a, p_a = stats.shapiro(df['Mat_A'])
print("Mat_A의 Shapiro-Wilk Test : statistic={}, p-value={}".format(round(statistic_a,3), round(p_a,3)))

Mat_A의 Shapiro-Wilk Test : statistic=0.906, p-value=0.256


##### Mat_B에 대한 정규성 검정  : 정규성 만족

Mat B의 분포의 경우, p-value가 0.973로 0.05보다 매우 크므로, 귀무가설을 기각하기 어렵다.  
따라서 Mat_B의 분포는 정규성을 만족한다고 할 근거가 충분하다.

In [53]:
statistic_b, p_b = stats.shapiro(df['Mat_B'])
print("Mat_B의 Shapiro-Wilk Test : statistic={}, p-value={}".format(round(statistic_b,3), round(p_b,3)))

Mat_B의 Shapiro-Wilk Test : statistic=0.973, p-value=0.917


### Q10.
#### 새로운 품종의 감자 20Box를 랜덤추출해 무게를 측정한 결과가 다음과 같다. (단, 모표준편차는 3.8로 알려짐) 이 자료로 부터 신품종 배추의 평균무게에 대한 95% 신뢰구간을 구하시오.  

모표준편차를 알고 있으므로, 정규분포를 이용하여 신뢰구간을 구할 수 있다.

In [56]:
lower, upper = stats.norm.interval(0.95, loc = df_potato.mean(), scale = 3.8/np.sqrt(len(df_potato)))
print("신뢰구간 : ({0}, {1})". format(lower.round(3), upper.round(3)))

신뢰구간 : ([17.835], [21.165])


위의 결과에 따라 신품종 배추의 평균무게는 (17.835, 21.165)의 범위에 있을 확률이 95% 정도 됨을 알 수 있다.

###  Q11.
#### 제품개발자는 새로 개발된 강종의 평균 인장 강도를 70kg/mm^2로 기대하고 있다. 이를 검증하기 위해 시험 생산되는 강종 중 임의로 20개의 코일을 대상으로 랜덤하게 시편을 채취하였고, 이 자료로 부터 모평균값의 95%로 신뢰구간을 추정해라.

##### - 정규성 검정 : 정규성 만족

In [66]:
df = [73,71,74,69,70,73,70,68,75,72,70,72,73,70,70,72,71,70,75,72]
statistic, p = stats.shapiro(df)
print("Mat_B의 Shapiro-Wilk Test : statistic={}, p-value={}".format(round(statistic,3), round(p,3)))

Mat_B의 Shapiro-Wilk Test : statistic=0.947, p-value=0.325


In [68]:
lower, upper = stats.t.interval(0.95, len(df)-1,  loc = np.mean(df), scale = stats.sem(df))
print("신뢰구간: ({0}, {1})" .format(lower.round(2), upper.round(2)))

신뢰구간: (70.6, 72.4)


모집단의 분포가 정규분포를 가정되지 않고, 모수에 대한 정보도 주어지지 않으며, 표본의 수도 매우 적다면, t-분포를 사용할 수 없는 것으로 알고 있다.  
그러나 Shapiro-Wilk 정규성 검정의 결과로, 모집단의 분포가 정규성을 만족한다고 주장이 가능하므로 t-분포를 통해 신뢰구간을 구할 수 있다.
따라서 위의 결과에 따라, 새로 개발된 강종의 평균 인장 강도는 (70.6, 72.4)의 범위에 있을 확률이 95% 정도 됨을 알 수 있다. 

### Q12.
#### 고객서비스센터의 고객만족도 평균은 78.0이다. 개선활동을 완료한 후 다음과 같은 10개의 고객만족도 데이터를 얻었을 때, 개선활동이 만족도를 변화시켰는가?

##### - 가설설정  
H0 : 고객만족도 평균은 78.0이다.  
H1 : 고객만족도 평균은 78.0이 아니다.

##### - 유의수준  
alpha = 0.05

##### - 정규성 검정 : 정규성 만족

In [63]:
df_satisfaction = [85.0, 79.0, 79.1, 79.9, 81.6, 78.6, 85.4, 83.4, 78.1, 79.2]
statistic, p = stats.shapiro(df_satisfaction)
print("statistic={}, p-value={}".format(round(statistic,3), round(p,3)))

statistic=0.847, p-value=0.054


##### - 1-sample t-test

In [64]:
t_result = stats.ttest_1samp(df_satisfaction, 78.0)
t,p = t_result.statistic.round(3), t_result.pvalue.round(3)
print('1-Sample t-test')
print("t 통계량 : {}".format(t))
print("p-value : {}".format(p))

1-Sample t-test
t 통계량 : 3.379
p-value : 0.008


##### 결과: 귀무가설 기각  
모집단의 분포가 정규분포를 가정되지 않고, 모수에 대한 정보도 주어지지 않으며, 표본의 수도 매우 적다면, t-분포를 사용할 수 없는 것으로 알고 있다.   
그러나 Shapiro - Wilk 정규성 검정의 p-value가 0.054로 귀무가설을 채택하여 정규성을 만족한다고 할 수 있으므로, t-분포를 이용한 평균 검정을 실시하였다.  그 결과, 0.008의 p-value 값으로 인해, 귀무가설을 기각할만한 충분한 근거를 얻을 수 있었고, 따라서 고객만족도의 평균은 78.0이라 주장하기 어렵다.

### Q13.
#### 제품을 A공급자와 B공급자로부터 제품을 받아 두께를 측정해서 다음의 데이터를 얻었다. 공급자 간에 모평균에 차이가 있다고 할 수 있는가?

##### - 가설설정  
H0 : A공급자의 제품의 평균과 B공급자의 제품의 평균에는 차이가 없다.(M1 = M2)  
H1 : A공급자의 제품의 평균과 B공급자의 제품의 평균에는 차이가 있다.(M1 != M2)

##### - 유의수준  
alpha = 0.05

In [76]:
df_suplies = pd.DataFrame({"A" : [1.883, 1.715, 1.799, 1.768, 1.711, 1.832, 1.427, 1.344],
                           "B" : [1.435, 1.572, 1.486, 1.511, 1.457, 1.548, 1.404, 1.800]})

##### - A의 정규성검정 결과 : 정규성 만족

In [77]:
statistic, p = stats.shapiro(df_suplies['A'])
print("A의 정규성검정 결과")
print("statistic={}, p-value={}".format(round(statistic,3), round(p,3)))

A의 정규성검정 결과
statistic=0.843, p-value=0.08


##### - B의 정규성검정 결과 : 정규성 만족

In [78]:
statistic, p = stats.shapiro(df_suplies['B'])
print("B의 정규성검정 결과")
print("statistic={}, p-value={}".format(round(statistic,3), round(p,3)))

B의 정규성검정 결과
statistic=0.839, p-value=0.074


##### - 등분산검정 결과 : 등분산으로 가정

In [79]:
stats.bartlett(df_suplies['A'], df_suplies['B'])

BartlettResult(statistic=1.2899966318541203, pvalue=0.25604857136527204)

##### - 2-Sample t-test

In [80]:
t_result = stats.ttest_ind(df_suplies['A'], df_suplies['B'])
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)
print('2-Sample t-test')
print('t-statistic:{}, p-value:{}'.format(round(t,3), round(p,3)))

2-Sample t-test
t-statistic:1.94, p-value:0.073


결과: 귀무가설 채택

Shapiro - Wilk 정규성 검정을 통해, A공급자의 제품. B공급자의 제품이 모두 정규성을 만족함을 확인할 수 있다.  
또한 두 집단의 정규성이 충족되어, Bartlett 등분산 검정을 시행하였고, 그 결과 두 집단의 분산은 같은 것으로 판단될 수 있음을 알 수 있었다.   
이러한 조건을 통한 2-sample t-test 검정 결과, 0.073의 p-value 값으로 인해 귀무가설을 기각할만한 충분한 근거를 얻지 못하였고, 따라서 두 공급자 A,B의 제품에 대한 평균에 차이가 없다고 주장할 근거가 충분하다.

### Q14.
#### 회사원 10명에 대해 영어 점수가 학원가기 전/후에 영어점수가 짝을 이루고 있다. 회사원들의 성적은 학원가기 전후에 대해 차이가 있는지 검정하라 

##### - 가설설정  
H0 : 회사원들의 영어 성적은 학원 가기 전후에 차이가 없다.(d1 - d2 =0)  
H1 : 회사원들의 영어 성적은 학원 가기 전후에 차이가 있다.(d1 - d2 != 0)

##### - 유의수준  
alpha = 0.05

In [81]:
df_before = pd.DataFrame({'before' : [720,589,780,648,720,589,780,648,780,648]})
df_after  = pd.DataFrame({'after'  : [710,580,787,712,750,600,782,670,790,680]})

##### - paired t -test

In [82]:
t_result = stats.ttest_rel(df_before, df_after)
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)
print("Paired t-test")
print("t:{}".format(t))
print("p:{}".format(p))

Paired t-test
t:[-2.266]
p:[0.05]


##### 결과: 귀무가설 기각
  
한 표본에 대하여, 처리 전후의 차이를 알아보기 위해서 paired t-test 검정을 시행하였다.  
paired t-test 검정 결과, 0.05의 p-value 값을 확인할 수 있었고 대립가설을 주장하여 기존의 가설을 바꾸고자 했기 때문에 귀무가설을 기각하는 것을 선택하였다.  
따라서 회사원들의 영어 성적은 학원 가기 전후에 차이가 없다고 주장할 근거가 충분하지 않다.

### Q15.
####  동전이 정상적인가를 검정하기 위해 100번을 던져본 결과, 앞면이 40번 나왔으며, 뒷면 60번 나왔다. 이 동전은 정상인가를 유의수준 5%로써 검정하라.  정상적인 동전을 던질 때, 앞면이 나올 확률은 50%이다.

##### - 가설설정  
H0 : 동전을 던졌을 떄, 앞면이 나올 확률은 0.5이다.(p = 0.5)  
H1 : 동전을 던졌을 떄, 앞면이 나올 확률은 0.5이 아니다.(p != 0.5)

##### - 유의수준  
alpha = 0.05

##### - 1-Proportion test

In [84]:
head = 60
n_obs = 100
value = 0.5
z,p = proportions_ztest(head, n_obs, value)
print("Proportiion test")
print('z:{0:0.3}'.format(round(z,3)))
print('p:{0:0.3}'.format(round(p,3)))

Proportiion test
z:2.04
p:0.041


##### 결과: 귀무가설 기각

모비율은 n, p가 커짐에 따라 정규분포로 근사한다. 따라서 정규분포를 통해 모비율에 대한 검정을 실시할 수 있다.   
위의 검정의 결과에 따라, 0.041의 p-value를 가지므로, 귀무가설을 기각할만한 근거가 충분하다고 할 수 있다.  
따라서 동전을 던졌을 때 앞면이 나올 확률은 0.5라고 할 근거가 충분하지 않으므로, 동전은 정상적이라고 보기 힘들다.

### Q16.
#### 동일한 제품을 생산하는 두 공장에서 불량률을 측정한 결과 아래와 같다. 두 공정의 불량률이 같다고 할 수 있는가?
1) 공장 A : N1 = 1200, X1 = 14  
2) 공장 B : N2 = 1200, X2 = 5

##### - 가설설정  
H0 : 두 공정의 불량률은 같다.(p1 = p2)  
H1 : 두 공정의 불량률은 같지 않다.(p1 != p2)

##### - 유의수준  
alpha = 0.05

##### - 2-Proportion test

In [85]:
count = np.array([14, 5])
n_obs = np.array([1200, 1200])
z,p = proportions_ztest(count, n_obs)
print("2 Proportiion test")
print('z:{0:0.3}'.format(round(z,3)))
print('p:{0:0.3}'.format(round(p,3)))

2 Proportiion test
z:2.07
p:0.038


##### 결과: 귀무가설 기각

위의 검정의 결과에 따라, 0.038의 p-value를 가지므로, 귀무가설을 기각할만한 근거가 충분하다고 할 수 있다.  
따라서 두 공정의 불량률이 같다고 주장할 근거가 충분하지 않다.

### Q17.
#### 노트북 컴퓨터의 6가지 제품(A, B, C, D, E, F)에 대하여 7가지 제품의 이미지를 중복 선택 할 수 있도록 32명의 소비자를 대상으로 설문조사한 Table이다. 각 제품에 대해 이미지 차이가 존재하겠는가?

##### - 가설설정  
H0 : 각 제품의 이미지에 대한 차이가 없다.  
H1 : 각 제품의 이미지에 대한 차이가 있다.

##### - 유의수준  
alpha = 0.05

In [101]:
df_survey = pd.DataFrame({'A' : [18,1,8,7,10,9,10], 'B' : [8,2,14,5,5,9,4],
                          'C' : [4,1,3,4,9,5,4],    'D' : [4,1,2,3,2,7,3],
                          'E' : [3,1,3,1,1,1,1],    'F' : [3,25,8,10,2,1,7]})
chi, pval, dof, expected = stats.chi2_contingency(df_survey)

print('chi-square test')
print('chisq : {0:0.3f}'.format(chi))
print('p:{0:0.3f}'.format(pval))
print('degree pd freedom : {}'.format(dof))

chi-square test
chisq : 98.274
p:0.000
degree pd freedom : 30


##### 결과 : 귀무가설 기각  
  
카이제곱 검정의 결과에 따라, p-value가 0에 근사하는 매우 작은 값을 나타내므로 귀무가설을 기각하기에 충분하다고 할 수 있다. 따라서 6개의 각 제품에 따른 이미지에 대한 차이가 없다고 주장할 근거가 충분하지 않다.

### Q18.
#### 철강의 청정도를 높이기 위해 제강공정에서 2차 정련을 실시하는데, 강종별 2차 정련시간의 차이가 있는지와 type별로 2차정련 시간에 차이가 있는지 검정하라

##### - 가설설정  
1) 강종에 따라  
H0 : 강종 별 정련시간의 평균에 차이가 없다.(m1 = m2 = m3 = m4)  
H1 : 적어도 하나의 강종 정련시간의 차이가 있다.(not H0)  
  
2) type에 따라  
H0 : type 별 정련시간의 평균에 차이가 없다.(m1 = m2)  
H1 : type 정련시간의 차이가 있다.(not H0)

##### - 유의수준  
alpha = 0.05

In [87]:
df_anova = pd.read_csv("../실습파일/BigData분석기초/ANOVA.csv")
df_anova.head()

time charge gangjong type
0  32.2      A        a    a
1  30.4      B        a    a
2  33.6      C        a    a
3  36.4      D        a    a
4  33.9      E        a    a

##### - gangjong, type의 unique value 확인

In [88]:
print(df_anova['gangjong'].unique())
print(df_anova['type'].unique())

['a' 'b' 'c' 'd']
['a' 'b']


In [89]:
df_gangjong_a = df_anova[df_anova['gangjong'] == 'a']['time']
df_gangjong_b = df_anova[df_anova['gangjong'] == 'b']['time']
df_gangjong_c = df_anova[df_anova['gangjong'] == 'c']['time']
df_gangjong_d = df_anova[df_anova['gangjong'] == 'd']['time']
df_type_a = df_anova[df_anova['type'] == 'a']['time']
df_type_b = df_anova[df_anova['type'] == 'b']['time']

##### - 강종별, type별 집단의  Shapiro-Wlik test  :  강종 4집단 , type 2집단 모두 정규성 만족

In [90]:
gangjong_list = [df_gangjong_a, df_gangjong_b, df_gangjong_c, df_gangjong_d]
type_list = [df_type_a, df_type_b]

In [95]:
print('강종별 집단의 Shapiro-wilk test')
print()
for gangjong_df, gangjong_type in zip(gangjong_list, ['a','b','c','d']):
    statistic, p = stats.shapiro(gangjong_df)
    print("gangjong", gangjong_type, "의 정규성 검정")
    print("statistic={}, p-value={}".format(round(statistic, 3), round(p, 3)))
    print()
print('**********************************************************************')
print('type별 집단의 Shapiro-wilk test')
print()
for type_df, type_type in zip(type_list, ['a','b']):
    statistic, p = stats.shapiro(type_df)
    print("type", type_type, "의 정규성 검정")
    print("statistic={}, p-value={}".format(round(statistic, 3), round(p, 3)))
    print()

강종별 집단의 Shapiro-wilk test

gangjong a 의 정규성 검정
statistic=0.988, p-value=0.984

gangjong b 의 정규성 검정
statistic=0.977, p-value=0.936

gangjong c 의 정규성 검정
statistic=0.893, p-value=0.334

gangjong d 의 정규성 검정
statistic=0.971, p-value=0.902

**********************************************************************
type별 집단의 Shapiro-wilk test

type a 의 정규성 검정
statistic=0.982, p-value=0.991

type b 의 정규성 검정
statistic=0.91, p-value=0.212



##### - 강종별 , type별 집단의 등분산검정 결과 : 강종 집단, type 집단 각각 등분산으로 가정

In [92]:
stats.bartlett(df_gangjong_a, df_gangjong_b, df_gangjong_c, df_gangjong_d)

BartlettResult(statistic=1.8729084232387967, pvalue=0.5991991808067023)

In [96]:
stats.bartlett(df_type_a, df_type_b)

BartlettResult(statistic=0.9529700642151103, pvalue=0.3289644435379086)

##### - one way anova

In [93]:
f_result = stats.f_oneway(df_gangjong_a, df_gangjong_b, df_gangjong_c, df_gangjong_d)
f, p = f_result.statistic.round(3), f_result.pvalue.round(3)
print("One-way ANOVA")
print("F 통계량 : {}".format(f))
print("p-value : {}".format(p))

One-way ANOVA
F 통계량 : 4.172
p-value : 0.019


In [98]:
f_result = stats.f_oneway(df_type_a, df_type_b)
f, p = f_result.statistic.round(3), f_result.pvalue.round(3)
print("One-way ANOVA")
print("F 통계량 : {}".format(f))
print("p-value : {}".format(p))

One-way ANOVA
F 통계량 : 9.654
p-value : 0.005


##### 결과: 두 가설 모두 귀무가설 기각

강종, type에 따른 각각의 2차 정련시간의 분포들이 정규성을 만족함을 확인할 수 있었다. 뿐만아니라 강종에 따른 4개의 집단 그리고 type에 따른 2개의 집단이 등분산으로 가정되어, 각각에 대한 anova 분석을 진행하였다.  
우선 강종에 따른 2차 정련시간의 집단의 평균에 차이는 적어도 한 집단은 있는 것으로 나타났으며, 다음으로 type에 따른 2차 정련시간의 집단에도 마찬가지로 적어도 한 집단은 차이가 있는 것으로 나타났다.  
추가적으로 type에 대한 2차 정련시간의 집단은 2개로 구성되어 anova분석 뿐아니라, 2-sample t-test도 가능함을 알 수 있다.

### Q19. 
#### 초코칩의 강도와 재료 반죽 온도 간에 관계를 분석하기 위한 상관분석을 하세요.

##### - 가설설정  
H0 : 초코칩의 강도와 재료 반죽 온도 간에 상관관계가 없다.(r = 0)  
H1 : 초코칩의 강도와 재료 반죽 온도 간에 상관관계가 있다.(r != 0)

In [99]:
strength = [37.6, 38.6, 37.2, 36.4, 38.6, 39, 37.2, 36.1, 35.9, 37.1,
            36.9, 37.5, 36.3, 38.1, 39, 36.9, 36.8, 37.6, 33, 33.5]
temperature = [14,15,14,16,17,14,17,16,15,16,14,16,18,13,15,17,14,16,20,21]

##### - Correlation Analysis

In [100]:
corr, pval = stats.pearsonr(strength, temperature)
print('Correlation Analysis')
print("corr:{0:0.3f}".format(corr))
print("p-value:{0:0.3f}".format(pval))

Correlation Analysis
corr:-0.723
p-value:0.000


##### 결과 : 귀무가설 기각  
  
상관계수 검정의 결과에 따라, p-value가 0에 근사하는 매우 작은 값을 나타내므로 귀무가설을 기각하기에 충분하다고 할 수 있다. 따라서 초코칩의 강도와 재료 반죽 온도 간에 상관관계가 있다고 할 근거가 충분하여, 두 변수 사이에 -0.723의 강한 음의 상관관계를 갖는다고 할 수 있다.